design-documents/Investigation-on-removal-of-dependency-on-Auto_Increment-Id-in-application.md

```diff
@@ -0,0 +1,443 @@
+## Investigate removing dependency on autoincrement id in application
+
+### Overview
+
+When trying to upgrade Magento to MySQL v8 an issue was encountered that relates to relying on the `AUTO_INCREMENT` of
+a table in several places. This document investigates on where this dependency is being used, why it is used and how
+difficult it will be to get rid of this dependency.
+Note that this document does not discuss the whether or not the dependency shall be removed but rather helps in making
+the decision.
+
+### List of places where `AUTO_INCREMENT` is used and for which purpose
+
+Currently the `AUTO_INCREMENT` is read from database in the following two methods:
+* `\Magento\ImportExport\Model\ResourceModel\Helper::getNextAutoincrement()`: Attempts to extract the `AUTO_INCREMENT`
+    from table status and returns it so that it can be used for predictive purposes.
+* `\Magento\Framework\Mview\View\Changelog::getVersion()`: Attempts to extract the `AUTO_INCREMENT` from table status,
+    decreases it by 1 and returns that. Actually returns the maximal version number that is currently in use.
+    This can probably be replaced by a query of the likes of `SELECT MAX(id) FROM table`.
```

Comment:
this is not an option, because this code is not thread-safe.

Reply:
@akaplya what is about LAST_INSERT_ID(); https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id  ?

Reply:
@akaplya Can you elaborate what you mean by "not thread safe". Could a situation occur where two processes try to generate a new materialized view for the same entity? Why is this not thread safe?

Reply:
@jean-bernard-valentaten last_increment_id unlike max(id)  will return last generated ID no matter to transaction and it's status.
So it is fine to use last_increment_id, btw you can find an example in SalesSequence module. but never rely on transactional data to determine the next sequential ID.

Reply:
For customers import, it is possible to utilise sequences, as we have for EE and Content Staging or Sales documents numbers.
https://www.percona.com/blog/2008/04/02/stored-function-to-generate-sequences/
In this case, instead of guessing a new ID import script may generate and a unique ID for each new customer or address.

To implement this solution in a consistent and compatible (with models) manner, you can add proper identifier handling into corresponding resource models. Logic should be pretty straightforward. If the resource model has to save an object with no ID and we cannot match this object with an existing record (by the unique key) in the database. For such case, the resource model can request a new ID for the object from a sequence.
See example from Sales module https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Sales/Model/ResourceModel/EntityAbstract.php#L136

Such a solution will be backward compatible, except the cases when somebody tries to insert customer data directly through the INSERT statement.

Reply:
@akaplya As far as I understand from the MySQL documentation `last_insert_id` is kept per connection only, thus if you have multiple connections, they will get different values when selecting `last_insert_id`. Also this is not supported by all RDBMS, so it would become an issue in case other RDBMS become a thing. I don't see how this is a reliable alternative.

Reply:
This is what we need, we need to generate a value that cannot be repeated by another session/process.
Call to a function last_insert_id, is implementation details. Which could be hidden by fancy sequence interface  Please see the example https://github.com/magento/magento2ce/blob/2.3-develop/app/code/Magento/SalesSequence/Model/Sequence.php#L83

Reply:
Agreed, I changed it.